// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

syntax = "proto2";
package kudu.ranger;

option java_package = "org.apache.kudu.ranger";

enum ActionPB {
  SCAN = 0;
  INSERT = 1;
  UPDATE = 2;
  DELETE = 3;
  ALTER = 4;
  CREATE = 5;
  DROP = 6;
  ALL = 7;
  METADATA = 8;
}

// Database/table pair
message TablePB {
  // Database name
  required string database = 1;

  // Table name.
  optional string table = 2;

  // Whether the database name was defaulted or actually written out.
  required bool default_database = 3;
}

// Authorization request
message AuthzRequestPB {
  // List of tables the action is to be performed on.
  repeated TablePB table = 1;

  // List of columns the action is to be performed on. Can only be used with a
  // single table.
  repeated string column = 2;

  // Action to be authorized on all tables or columns.
  required ActionPB action = 3;

  // User performing the action
  required string user = 4;
}

// Authorization response
message AuthzResponsePB {
  // List of tables the user is authorized to perform the action on.
  repeated TablePB table = 1;

  // List of columns. Can only be used with a single table.
  repeated string column = 2;
}
